neuralpp/symbolic/expression.py

```diff
@@ -164,13 +184,13 @@ def new_variable(cls, name: str, type_: ExpressionType) -> Variable:
 
     @classmethod
     @abstractmethod
-    def new_function_application(cls, function: Expression, arguments: List[Expression]) -> FunctionApplication:
+    def new_function_application(cls, function: Expression, arguments: List[Expression]) -> Expression:
         pass
 
     @classmethod
     @abstractmethod
     def new_quantifier_expression(cls, operation: Constant, index: Variable, constrain: Expression, body: Expression,
-                                  ) -> QuantifierExpression:
+                                  ) -> Expression:
```

Comment:
No.

https://github.com/rodrigodesalvobraz/neuralpp/pull/29/files/becc5084c2414eb21a2f88a3815b71e2565082e0#diff-12ba35f9b7c823ba859085be51887691ff64f8d9eefb76a83076b1572b2c4282R78-R82

Reply:
Similarly `new_function_application()` is not guaranteed to return `FunctionApplication`. E.g., `SymPyExpression.new_function_application()` is allowed to simplify the result in some circumstances (e.g., `(and False b)`).

Reply:
This is how I did it in my first implementation, and later regretted it. Today I think it is better to separate the construction of expressions and their simplification, which here are always done together. I think it is better because it is more regular: building a quantifier expressions always produces a quantifier expression, making it easier to write algorithms. It also makes typing more precise (as illustrated above). If the user wants things simplified, they can always do so.

So, if you think this is easy enough to separate, that would be good. If it is too much work, it is ok to leave it.

Reply:
This is almost impossible to implement, as I explained in the reply below.